Help Cockroach Labs Flow

maxlang opened the pull request:
While it isn't 100% polished, this is a first pass at the flow that encourages Cockroach DB users to share their usage information.

**New support/opt-in page:**
<img width="662" alt="screen shot 2016-02-16 at 3 40 35 am" src="https://cloud.githubusercontent.com/assets/3691886/13071111/42c00ee2-d45f-11e5-8667-b04f33c3e505.png">

**Opt in Banner**
<img width="601" alt="screen shot 2016-02-16 at 3 41 08 am" src="https://cloud.githubusercontent.com/assets/3691886/13071106/42be2b04-d45f-11e5-971b-4ffc9a33c484.png">

**Signup Modal**
<img width="588" alt="screen shot 2016-02-16 at 3 41 15 am" src="https://cloud.githubusercontent.com/assets/3691886/13071107/42be9f1c-d45f-11e5-9caa-c82707b49654.png">

**Thanks Modal**
<img width="563" alt="screen shot 2016-02-16 at 3 41 36 am" src="https://cloud.githubusercontent.com/assets/3691886/13071110/42bfb578-d45f-11e5-9eca-86d098eb2626.png">

**Filled in Information**
<img width="589" alt="screen shot 2016-02-16 at 3 41 47 am" src="https://cloud.githubusercontent.com/assets/3691886/13071108/42bf589e-d45f-11e5-85a1-ceb3635a7882.png">

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4397)

<!-- Reviewable:end -->

Commit:
help us flow

knz commented:
Hi Max I think you should revisit the split of the name field into separate "first name" and "last name".
See http://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/

For example the distinction First/Last does not work well for Chinese people. There are places where people only have one name that matters. In Vietnam they have 3 used for different purposes. 

Really a single field "Full name" would probably work better. Don't forget to accept any unicode sequence in there :)

knz commented:
Also having a single field "Name" would go a long way towards "make databases easier". One less field to fill in / scroll through etc.

JackKrupansky commented:
Very minor nit: Github s.b. GitHub.

jseldess commented:
Looking great, @maxlang. Just a few minor suggestions and thoughts:

#### Opt-in banner

Remove period after "Check for updates and help us improve".

#### Sign up modal 

I like the first sentence on the the support page (before opt-in). What if we change the opt-in banner text a little:

> CockroachDB is in beta, and we're working diligently to make it better. Opt in to share basic, anonymous usage statistics, report bugs and feedback directly from the UI, and get notified when new versions of CockroachDB are available.    

Also wonder if we shouldn't just shorten the modal title to "Help Cockroach Labs"?

#### Support page (before opt-in)

What if we mirror the new text of the sign up modal, with the addition of the link to GitHub?: 

> CockroachDB is in beta, and we're working diligently to make it better. Opt in to share basic anonymous usage statistics, report bugs and feedback directly from the UI, and get notified when new versions of CockroachDB are available, or just hit us up on GitHub (link).
  
#### Support page (after opt-in)

After signing up, shouldn't there be a way to submit issues directly from the UI? If not, our opt-in text is misleading... 

Also, shouldn't users have a way to opt out from this page?

cc: @jess-edwards

RaduBerinde commented:
On the "Help Make CockroachDB Better" dialog I would add more information explaining exactly what gets sent and what doesn't. I assume no table data is sent, but are we sending schema information? If yes, are table and column names sent as-is or anonymized? These should be mentioned. If it's too much to list in the dialog, we should add a link to a page that will explain these details.

Also,space after "better."

maxlang commented:
@knz I agree for all the reasons given and actually originally had a "fullname/name" field, but the analytics service we were looking at wanted first/last names. I'll talk to @jess-edwards since I think this is preferable.

knz commented:
Wait what? Because our potential external dependencies are working backwards we should lower our own standards? 
I'd day do the Right Thing upfront, and only then select the partners who are ready to work with you this way. 

Tail wagging the dog and all that...
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

bdarnell commented:
Could we make name optional if company is provided? If I were filling out this form for a company-owned cluster I'd want to put a team alias (`db-team@example.com`) in the email field instead of an individual's address, in which case there's no useful "name" to be provided.

jess-edwards commented:
@maxlang 

Copy feedback: 

### New Support/opt-in page: 
@jseldess I agree that the Sign-Up Modal and "Opt-In page" should have identical language. 
cut: "or just hit us up on GitHub." It's not relevant to this list. 

### Opt-in Banner: 
@jseldess I don't think "Help Cockroach Labs" gives enough information or incentive about why a person should opt in. "Check for updates and help us improve" is more clear. 

### Support page (before opt-in)
@jseldess I agree. There should be a way to submit bugs here. If a user does not want to opt-in in order to file a bug, we should provide an alternative. 

@maxlang What does the "report a bug" page look like and what is that workflow for a user who has opted in vs a user who hasn't?

jseldess commented:
Thanks, @jess-edwards. Only one thing I want to clarify. Wasn't suggesting to change to "Help Cockroach Labs" on the opt-in banner (I agree that what's there is good) but rather the sign up modal. Thoughts?

maxlang commented:
@jess-edwards @jseldess Good points, I'll remove language about submitting bugs and just include a link to the Google group instead - I think that's as far as we want to go for now since we're not handling bug reports from the UI in a special way.

mrtracy commented on ui/ts/components/navbar.ts:
```diff
@@ -35,20 +35,29 @@ module Components {
-          )
-      ));
+        _.map(args.ts.targets(), function generateLinks(t: Target): MithrilVirtualElement {
+        let linkAttrs: any = {
```
Can you put a comment here about handling relative URLS vs absolute URLS

maxlang replied:
done

mrtracy commented on ui/ts/pages/helpus.ts:
```diff
@@ -0,0 +1,215 @@
+        show: boolean = false;
+
+        // The closeOnEscape function, but bound to the correct `this`.
+        closeOnEscapeBound: (e: KeyboardEvent) => void;
```
you can just declare this in the body of the class as 
```typescript
closeOnEscape = (e: keyboardEvent) => {
   // body of function
} 
```

There's no need to have both `closeOnEscape` and `closeOnEscapeBound`

mrtracy commented on ui/ts/pages/helpus.ts:
```diff
@@ -0,0 +1,215 @@
+
+        if (ctrl.show) {
+          // Banner
+          if (ctrl.stage === Stage.Banner) {
```
Use a `switch` statement here, it will be much easier to read.

maxlang replied:
To be honest, I've always stayed away from switch fallthroughs because they're kind of error prone (as mentioned here in Javascript the Good Parts http://archive.oreilly.com/pub/a/javascript/excerpts/javascript-good-parts/bad-parts.html) but this is kind of the perfect place for it, so I guess I'll cave.

mrtracy commented on ui/ts/pages/settings.ts:
```diff
@@ -0,0 +1,116 @@
+      import UserOptIn = Models.HelpUs.UserOptIn;
+
+      class HelpUsController {
+        userData: UserOptIn = new UserOptIn();
```
It should be fine to make this a global. Right now, the modal and the page get a different copy of the UserOptIn().

maxlang replied:
Sounds good.

mrtracy commented on ui/ts/pages/settings.ts:
```diff
@@ -0,0 +1,116 @@
+
+      class HelpUsController {
+        userData: UserOptIn = new UserOptIn();
+        signedUp: boolean = false;
```
Shouldn't this just be part of the UserOptIn() model?

maxlang replied:
Makes sense.

mrtracy commented on ui/ts/pages/helpus.ts:
```diff
@@ -0,0 +1,215 @@
+          this.closeOnEscapeBound = this.closeOnEscape.bind(this);
+          document.addEventListener("keydown", this.closeOnEscapeBound);
+
+          m.sync([
```
I don't quite understand this; why does showHelpUs() need to be a promise?  You do not need to conflate the "show-help-us" URL parameter with anything in the database.

The UserOptIn() should be a "global" of sorts, there's only one per user and it might be used by the registration page. There's also no reason not to sync it on every load.  

The logic here on the modal dialog should be in the view, and it should be "IF the help-us flag is in the URL, AND the global UserOptIn() has been synced and does not indicate that the user has opted out, we can reload."

maxlang replied:
Yeah "show help us" is making the full check here - checking if it's filled in or dismissed or if the flag isn't present. But since we're already pulling in the data for the ```UserOptIn``` object, we might as well just use that. I'll simplify it.

maxlang commented:
@knz @bdarnell @jess-edwards 

I'm going to create an issue to discuss first/last name vs company name vs fullname.

mrtracy commented on ui/ts/pages/helpus.ts:
```diff
@@ -0,0 +1,215 @@
+            ]);
+          // Help Us Form
+          } else if (ctrl.stage === Stage.Modal || ctrl.stage === Stage.ThanksSize || ctrl.stage === Stage.ModalStart || ctrl.stage === Stage.ModalClose) {
+            return m(".modal-container" + containerClass, [
```
As mentioned in person, if you want to use the controller to bind and unbind the "closeOnEscape" function, the appropriate thing to do is to make the code in this section of the "if" statement (which occurs when a modal dialog is actually being displayed) into its own SubComponent.  That subComponent would be instantiated whenever the modal is displayed, and would be _unloaded_ when the modal is closed, making it appropriate for binding the "closeOnEscape".

mrtracy commented on ui/ts/models/helpus.ts:
```diff
@@ -0,0 +1,110 @@
+      updates: boolean = false; // Did the user sign up for product/feature updates
+    }
+
+    export class UserOptIn {
```
As discussed, this particular model should likely be accessed as a singleton.  Mithril has no specific recommendations around this, so I would just create a method `GetUserOptIn()` which returns a singleton instance of `UserOptIn`, and use that in place of `new UserOptIn()` in your page controllers.

maxlang replied:
I assigned a singleton variable in the module and exported it vs the function, but I think the addresses your comment.

mrtracy commented on ui/ts/models/helpus.ts:
```diff
@@ -0,0 +1,110 @@
+            return !target.checkValidity();
+          },
+        };
+      }
```
I would add some additional methods to this model, such as `userDismissed(): boolean`, which will be useful in some of your controllers.

mrtracy commented on sql/system.go:
```diff
@@ -150,6 +167,7 @@ func addSystemDatabaseToSchema(target *MetadataSchema) {
 	target.AddDescriptor(keys.SystemDatabaseID, &descriptorTable)
 	target.AddDescriptor(keys.SystemDatabaseID, &usersTable)
 	target.AddDescriptor(keys.SystemDatabaseID, &zonesTable)
+	target.AddDescriptor(keys.SystemDatabaseID, &ReportingTable)
```
Make sure you run all of the sql tests; I believe you may have to modify some expected data in a few of the SQL logic tests (specifically the file "sql/testdata/system").

maxlang replied:
Thanks

mrtracy commented on sql/system.go:
```diff
@@ -85,6 +98,9 @@ var (
 	zonesTable = createSystemTable(keys.ZonesTableID, zonesTableSchema)
 
+	// ReportingTable is the descriptor for the cluster reporting opt-in table.
+	ReportingTable = createSystemTable(keys.ReportingTableID, reportingTableSchema)
```
This does not need to be exported, make it `reportingTable`

maxlang replied:
fixed

maxlang commented:
@mrtracy made code fixes (no copy fixes just yet)

petermattis commented on sql/system.go:
```diff
@@ -61,6 +61,19 @@ CREATE TABLE system.zones (
+  company     STRING,
+  updates     BOOLEAN,
+  lastUpdated TIMESTAMP
+);`
```
Is this table ever going to have more than a single row? Breaking out all of the fields like this is classic schema design, but it makes adding new fields more irritating. As an alternative, did you consider having a `system.ui` table which looks like:

```
CREATE TABLE system.ui (
  key STRING PRIMARY KEY,
  json BLOB
)
```

The "reporting" data would be stored under the `"reporting"` key and all of the details would be in the `json` column. New details could be added without adjusting the schema. And then this `ui` table could be used to store other `ui` state as well.

I wouldn't necessarily change anything in this PR, but we should revisit when @cuongdo provides JSON endpoints in #4298.

maxlang replied:
This makes sense and seems better than what I'm doing. I'd maybe just add a user column for when we have per-user settings (it would be null for reporting.)

petermattis replied:
If you have per-user settings, the username could be contained within the key.

maxlang commented:
@jess-edwards @jseldess copy changes:

I changed the support page to "Help Us" and changed the language to reflect the fact that it's no longer a support page (though it may be one again in the future.)
![screen shot 2016-02-22 at 11 40 29 am](https://cloud.githubusercontent.com/assets/3691886/13225630/a760d3da-d95b-11e5-9beb-1be7b6545182.png)

Here's the same language in the modal.
![screen shot 2016-02-22 at 11 42 09 am](https://cloud.githubusercontent.com/assets/3691886/13225658/bfdd8160-d95b-11e5-858f-f29e032ef648.png)

Note that opting in is now an explicit checkbox and if you don't check it you can't submit:
![screen shot 2016-02-22 at 11 42 16 am](https://cloud.githubusercontent.com/assets/3691886/13225681/dca77b84-d95b-11e5-9e96-c4b74d6f9826.png)

Once you've opted in to email updates, that option disappears (since we don't want to administer your email preferences from the cluster admin UI.)
![screen shot 2016-02-22 at 1 16 34 pm](https://cloud.githubusercontent.com/assets/3691886/13227842/a90aaf84-d966-11e5-9077-a967f2146b55.png)

maxlang commented:
TODO: the privacy policy link doesn't work currently.

mrtracy commented:
Code is LGTM

jess-edwards commented:
@maxlang I think you should cut "Thanks for signing up!" Otherwise LGTM.

maxlang commented:
@jess-edwards should I get rid of it altogether? Right now I have a modal telling you how you can change your settings, but I can just skip that.

<img width="541" alt="screen shot 2016-02-23 at 12 12 27 am" src="https://cloud.githubusercontent.com/assets/3691886/13242396/3bf12b1a-d9c2-11e5-97c3-ea1df9cd340c.png">

maxlang commented:
And I removed the thanks language from here:

<img width="714" alt="screen shot 2016-02-23 at 12 41 26 am" src="https://cloud.githubusercontent.com/assets/3691886/13242773/58aa89aa-d9c6-11e5-91f2-404fc7465abe.png">

jess-edwards commented:
@maxlang I think it's appropriate in the modal. Keep it there.

jseldess commented:
@maxlang, very minor: The "send me updates" checkboxes on the **Usage Reporting page** and **signup modal** are different. On the former, it's "Send me updates about Cockroach". On the latter, it's "Send me product and feature updates."

I like explicitly mentioning that these updates will be product-related, so I prefer the latter. Thoughts, @jess-edwards? If we go with that, can you remove the period?

jess-edwards commented:
@jseldess @maxlang I agree with Jesse. Both should be "Send me product and feature updates."

maxlang commented:
@jseldess good catch

jess-edwards commented:
@maxlang The HubSpot integration on this page is broken. 

1. Where is the data going right now? Another way to ask that is - is anything being collected right now / is the "help us" form live and working? 
2. Can the HubSpot integration be fixed? 

@kuanluo
